test/functional/setup/ksp_app.c

```diff
@@ -0,0 +1,300 @@
+
+ ******************************************************************************/
+
+#define KSPNAME             L"MyCustomKSPName"
+#define PROV_ENCRYPT_KEY    "LPKCWVD07N3RG98J0MBLG4H2"  /* this can be any character string */
+
+#include <stdio.h>
+#include <stdlib.h>
+#ifdef _WIN32
+#include <windows.h>
+#else
+#define __stdcall
+#include <dlfcn.h>
+#endif
+#include <sql.h>
+#include <sqlext.h>
+#include "msodbcsql.h"
+
+enum job {
+    set_up = 0,
+    clean_up = 1
+};
+
+/* Convenience functions */
+
+int checkRC(SQLRETURN rc, char *msg, int ret, SQLHANDLE h, SQLSMALLINT ht) {
+    if (rc == SQL_ERROR) {
+        fprintf(stderr, "Error occurred upon %s\n", msg);
+        if (h) {
+            SQLSMALLINT i = 0;
+            SQLSMALLINT outlen = 0;
+            char errmsg[1024];
+            while ((rc = SQLGetDiagField(
+                ht, h, ++i, SQL_DIAG_MESSAGE_TEXT, errmsg, sizeof(errmsg), &outlen)) == SQL_SUCCESS
+                || rc == SQL_SUCCESS_WITH_INFO) {
+                fprintf(stderr, "Err#%d: %s\n", i, errmsg);
+            }
+        }
+        if (ret)
+            exit(ret);
+        return 0;
+    }
+    else if (rc == SQL_SUCCESS_WITH_INFO && h) {
+        SQLSMALLINT i = 0;
+        SQLSMALLINT outlen = 0;
+        char errmsg[1024];
+        printf("Success with info for %s:\n", msg);
+        while ((rc = SQLGetDiagField(
+            ht, h, ++i, SQL_DIAG_MESSAGE_TEXT, errmsg, sizeof(errmsg), &outlen)) == SQL_SUCCESS
+            || rc == SQL_SUCCESS_WITH_INFO) {
+            fprintf(stderr, "Msg#%d: %s\n", i, errmsg);
+        }
+    }
+    return 1;
+}
+
+void postKspError(CEKEYSTORECONTEXT *ctx, const wchar_t *msg, ...) {
+    if (msg > (wchar_t*)65535)
+        wprintf(L"Provider emitted message: %s\n", msg);
+    else
+        wprintf(L"Provider emitted message ID %d\n", msg);
+}
+
+int setKSPLibrary(SQLHSTMT stmt) {
+    unsigned char CEK[32];
+    unsigned char *ECEK;
+    unsigned short ECEKlen;
+    int i;
+#ifdef _WIN32
+    HMODULE hProvLib;
+#else
+    void *hProvLib;
+#endif
+    CEKEYSTORECONTEXT ctx = {0};
+    CEKEYSTOREPROVIDER **ppKsp, *pKsp;
+    int(__stdcall *pEncryptCEK)(CEKEYSTORECONTEXT *, errFunc *, unsigned char *, unsigned short, unsigned char **, unsigned short *);
+    
+    /* Load the provider library */
+#ifdef _WIN32
+    if (!(hProvLib = LoadLibrary("myKSP.dll"))) {
+#else
+    if (!(hProvLib = dlopen("./myKSP.so", RTLD_NOW))) {
+#endif
+        fprintf(stderr, "Error loading KSP library\n");
+        return 2;
+    }
+#ifdef _WIN32
+    if (!(ppKsp = (CEKEYSTOREPROVIDER**)GetProcAddress(hProvLib, "CEKeystoreProvider"))) {
+#else
+    if (!(ppKsp = (CEKEYSTOREPROVIDER**)dlsym(hProvLib, "CEKeystoreProvider"))) {
+#endif
+        fprintf(stderr, "The export CEKeystoreProvider was not found in the KSP library\n");
+        return 3;
+    }
+    while (pKsp = *ppKsp++) {
+        if (!memcmp(KSPNAME, pKsp->Name, sizeof(KSPNAME)))
+            goto FoundProv;
```

Comment:
goto? really?

Reply:
I basically got this from the sample code, but of course I can avoid using `goto` in this case.